Tool for preparing Fedora Image on the chromebook

dorindabassey opened the pull request:
This tool reuses some of the functions for creating ChromeOS bootable partitions, 
mounts the ROOTFS partition, prepares the Fedora rootfs, generates a FIT image using the kernel https://download.copr.fedorainfracloud.org/results/eballetbo/fedora/fedora-35-aarch64/03507242-kernel/ for the kernel partition, signs the kernel and creates a bootable media for the chromebook.

Signed-off-by: Dorinda Bassey <dbassey@redhat.com>

eballetbo commented on chromebook-setup.sh:
```diff
@@ -346,20 +379,29 @@ create_fit_image()
          mkimage -D "-I dts -O dtb -p 2048" -f auto -A ${ARCH} -O linux -T kernel -C $compression -a 0 \
-                 -d arch/${ARCH}/boot/$kernel $dtbs \
-                 kernel.itb
+                 -d arch/${ARCH}/boot/$kernel $dtbs kernel.itb
```
This change is not needed. In fact nothing changes, so just leave the old style

dorindabassey replied:
Ok, noted

eballetbo commented on chromebook-setup.sh:
```diff
@@ -410,7 +452,9 @@ cmd_format_storage()
     wait_for_partitions_to_appear
     find_partitions_by_id
 
+    sudo mkfs.ext4 -L kernel "$CB_SETUP_STORAGE1"
```
I am not sure why is needed this change. In fact the kernel partition is not a ext4 partition but a raw partition. Probably is overwritten when the kernel is flashed. Did you try to remove this line?

dorindabassey replied:
ok, I'll remove it then.

eballetbo commented on chromebook-setup.sh:
```diff
@@ -410,7 +452,9 @@ cmd_format_storage()
 
+    sudo mkfs.ext4 -L kernel "$CB_SETUP_STORAGE1"
     sudo mkfs.ext4 -L ROOT-A "$CB_SETUP_STORAGE2"
+    sudo cgpt add -i 1 -T 5 -P 10 -t kernel -l kernel "$CB_SETUP_STORAGE"
```
This should not be needed. Is not working if you remove this line and the above that you added?

eballetbo commented on chromebook-setup.sh:
```diff
@@ -704,6 +841,7 @@ ensure_command mkfs.ext4 e2fsprogs
 ensure_command mkimage u-boot-tools
 ensure_command udisksctl udisks2
 ensure_command vbutil_kernel vboot-utils
+ensure_command cgpt vboot-utils
```
vboot-utils is checked just in the above line, and cgpt is part of vboot-utils. So this line is not needed.

dorindabassey replied:
Ah, Ok Thanks

eballetbo commented on chromebook-setup.sh:
```diff
@@ -650,6 +694,99 @@ cmd_eject_storage()
+    local kernel_version="5.16.2-0.test.fc35.aarch64"
+
+    # Download a known kernel that works for Chromebooks
+    [ -f kernel-core-$kernel_version.rpm ] || wget https://download.copr.fedorainfracloud.org/results/eballetbo/fedora/fedora-35-aarch64/03507242-kernel/kernel-core-$kernel_version.rpm
```
Can we use `curl -OL`? So we don't introduce a new tool

eballetbo commented on chromebook-setup.sh:
```diff
@@ -650,6 +694,99 @@ cmd_eject_storage()
+
+    # Download a known kernel that works for Chromebooks
+    [ -f kernel-core-$kernel_version.rpm ] || wget https://download.copr.fedorainfracloud.org/results/eballetbo/fedora/fedora-35-aarch64/03507242-kernel/kernel-core-$kernel_version.rpm
+    [ -f kernel-modules-$kernel_version.rpm ] || wget https://download.copr.fedorainfracloud.org/results/eballetbo/fedora/fedora-35-aarch64/03507242-kernel/kernel-modules-$kernel_version.rpm
```
Same here. Can we use `curl -OL`? So we don't introduce a new tool

eballetbo commented on chromebook-setup.sh:
```diff
@@ -205,6 +229,15 @@ if [ -z "$CB_KERNEL_PATH" ]; then
+    curl -OL $GETFEDORA
+
+    echo "Decompress .xz image"
+    sudo unxz Fedora-Workstation-36-1.5.aarch64.raw.xz
```
We can avoid a hardcoded file name here. I.e $(basename $GETFEDORA). So when we change the value of the GETFEDORA in chromebook-config.sh we don't need to update that line.

dorindabassey replied:
agreed

eballetbo commented on chromebook-setup.sh:
```diff
@@ -205,6 +229,15 @@ if [ -z "$CB_KERNEL_PATH" ]; then
+
+    echo "Decompress .xz image"
+    sudo unxz Fedora-Workstation-36-1.5.aarch64.raw.xz
+    IMAGE="Fedora-Workstation-36-1.5.aarch64.raw"
```
Same here. $(basename -s .xz $GETFEDORA)

eballetbo commented on chromebook-setup.sh:
```diff
@@ -650,6 +694,99 @@ cmd_eject_storage()
     echo "All done."
 }
 
+cmd_setup_fedora_rootfs()
```
Can we add a comment like this just before these functions?

+# -----------------------------------------------------------------------------
+# Experimental: Create Fedora images for Chromebooks

eballetbo reviewed (changes requested):
Many thanks for your contribution Dorinda. Some comments below.

eballetbo commented:
Hi Dorinda, I tried the following command
```
./chromebook-setup.sh deploy_fedora --architecture=arm64 --distro=fedora --storage=/dev/sda
```
And I was expecting the script to download the default image and deploy to the pendrive, but doesn't work. Could you double check? 

Also, as improvement, I'd like to avoid to pass the distro paramater too, it is implicid when you run deploy_fedora, you can set the variable there, so we can simply do:

```
./chromebook-setup.sh deploy_fedora --architecture=arm64 --storage=/dev/sda
```

dorindabassey commented:
Hi Enric, yes you're right I just made fixes to resolve the download issue and improved the script to avoid passing distro parameter.

eballetbo commented on chromebook-setup.sh:
```diff
@@ -655,6 +689,103 @@ cmd_eject_storage()
+        IMAGE=$(basename -s .xz $GETFEDORA)
+    fi
+
+    fedora_fit=true
```
I think you can use CB_DISTRO here and set to 'fedora' so you don't need to add a new variable.

eballetbo commented on chromebook-setup.sh:
```diff
@@ -655,6 +689,103 @@ cmd_eject_storage()
+cmd_deploy_fedora()
+{
+    if [ ! -f "$IMAGE" ] && [ "$IMAGE" != "" ] || [ -z "$IMAGE" ]; then
+        echo "Error: $IMAGE not found! Downloading the default fedora image."
```
I'd remove "Error $IMAGE not found!" and just put "Downloading the default fedora image" as if you don't pass any image is confusing to see an error. Not passing an image is just fine.

eballetbo commented on chromebook-setup.sh:
```diff
@@ -655,6 +689,103 @@ cmd_eject_storage()
+
+cmd_deploy_fedora()
+{
+    if [ ! -f "$IMAGE" ] && [ "$IMAGE" != "" ] || [ -z "$IMAGE" ]; then
```
This if doesn't work as you expect. For example if you try twice

```
./chromebook-setup.sh deploy_fedora --architecture=arm64 --storage=/dev/sda
```
The image always try to download., even you already have it downloaded. This is because the image is uncompressed and stored as Fedora-Workstation-36-1.5.aarch64.raw and you're checking for the compressed name Fedora-Workstation-36-1.5.aarch64.raw.xz which doesn't exist.

dorindabassey replied:
Thank you for the review, I added a fix for optimization there.

eballetbo reviewed (changes requested):
Hi Dorinda, I think we're very close. Just a couple of reviews that I think would be worth to solve before merge. Thanks.

Commit:
Tool for preparing Fedora Image on the chromebook

This tool reuses some of the functions for creating ChromeOS bootable partitions,
mounts the ROOTFS partition, prepares the Fedora rootfs, generates a FIT image using the kernel https://download.copr.fedorainfracloud.org/results/eballetbo/fedora/fedora-35-aarch64/03507242-kernel/ for the kernel partition, signs the kernel and creates a bootable media for the chromebook.

v3:
- cmd_deploy_fedora: added a check for Image if the IMAGE var is being set.
- cmd_deploy_fedora: if the --image var is not being used, check if the default uncompressed image is unavailable, then download and decompress the image

Signed-off-by: Dorinda Bassey <dbassey@redhat.com>

eballetbo commented:
I tweaked a bit the commit message to adapt the style to the commits already done, but the PR is done and merged. Thanks for your contribution!
